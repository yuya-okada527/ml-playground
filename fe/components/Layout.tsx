import React, { ReactNode } from "react";
import Head from "next/head";
import AppBar from "@material-ui/core/AppBar";
import HomeIcon from "@material-ui/icons/Home";
import InfoIcon from "@material-ui/icons/Info";
import {
  Link,
  Grid,
  createStyles,
  makeStyles,
  Theme,
  Typography,
  List,
  ListItem,
} from "@material-ui/core";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    title: {
      margin: theme.spacing(2),
    },
    offset: {
      marginTop: theme.spacing(13),
    },
    sidebarIcon: {
      margin: theme.spacing(1),
    },
    sidebarText: {
      marginTop: theme.spacing(0.5),
    },
    footer: {
      width: "100%",
      position: "absolute",
      bottom: 0,
    },
    footerListItem: {
      paddingTop: theme.spacing(0.5),
      paddingBottom: theme.spacing(0),
    },
  })
);

type Props = {
  children?: ReactNode;
  title?: string;
};

const Layout = ({ children, title = "This is the default title" }: Props) => {
  const classes = useStyles();
  return (
    <div>
      <Head>
        <title>{title}</title>
        <meta charSet="utf-8" />
        <meta name="viewport" content="initial-scale=1.0, width=device-width" />
        <link
          rel="stylesheet"
          href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"
        />
      </Head>
      <header>
        <AppBar>
          <Typography variant="h4" component="h1" className={classes.title}>
            Movie Recommender
          </Typography>
        </AppBar>
      </header>
      <div className={classes.offset} />
      <Grid container>
        <Grid item xs={2}>
          <List>
            <ListItem>
              <Link href="/">
                <Grid container>
                  <HomeIcon className={classes.sidebarIcon} />
                  <Typography className={classes.sidebarText} variant="h6">
                    Home
                  </Typography>
                </Grid>
              </Link>
            </ListItem>
            <ListItem>
              <Link href="/about">
                <Grid container>
                  <InfoIcon className={classes.sidebarIcon} />
                  <Typography className={classes.sidebarText} variant="h6">
                    About
                  </Typography>
                </Grid>
              </Link>
            </ListItem>
          </List>
        </Grid>
        <Grid item xs={10}>
          {children}
        </Grid>
      </Grid>

      <footer className={classes.footer}>
        <hr />
        <List>
          <ListItem className={classes.footerListItem}>
            Contact: hoge@gmail.com
          </ListItem>
          <ListItem className={classes.footerListItem}>
            Github: https://github.com/hoge
          </ListItem>
        </List>
      </footer>
    </div>
  );
};

export default Layout;
