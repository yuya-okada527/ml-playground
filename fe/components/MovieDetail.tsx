import React from "react";
import {
  Card,
  CardMedia,
  CardContent,
  createStyles,
  makeStyles,
  Theme,
  Typography,
} from "@material-ui/core";
import { Movie } from "../interfaces";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    cardRoot: {
      display: "flex",
      marginTop: theme.spacing(5),
      height: "450px",
      width: "98%",
    },
    posterImage: {
      flex: 1.2,
    },
    movieDescription: {
      flex: 2,
    },
    movieTitle: {
      marginBottom: theme.spacing(2),
    },
  })
);

const MAX_OVERVIEW_LENGTH = 450;

const cutLongOverview = (overview?: string) => {
  if (!overview) {
    return overview;
  }

  if (overview.length > MAX_OVERVIEW_LENGTH) {
    return overview.substr(0, MAX_OVERVIEW_LENGTH) + "...";
  }

  return overview;
};

type MovieDetailProps = {
  movie_detail: Movie;
};

const MovieDetail = ({ movie_detail }: MovieDetailProps) => {
  const classes = useStyles();
  console.log(JSON.stringify(movie_detail));
  return (
    <Card className={classes.cardRoot}>
      <CardMedia
        className={classes.posterImage}
        image={movie_detail.poster_url}
      />
      <CardContent className={classes.movieDescription}>
        <Typography variant="h6" className={classes.movieTitle}>
          {movie_detail.japanese_title}
          {movie_detail.release_year
            ? " (" + movie_detail.release_year + ") "
            : ""}
        </Typography>
        <Typography variant="body1">
          {cutLongOverview(movie_detail.overview)}
        </Typography>
      </CardContent>
    </Card>
  );
};

export default MovieDetail;
