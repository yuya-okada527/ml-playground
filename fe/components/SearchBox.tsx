import React, { ChangeEvent } from "react";
import {
  Button,
  createStyles,
  makeStyles,
  TextField,
  Theme,
  Typography,
  Grid,
} from "@material-ui/core";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    searchButton: {
      marginTop: theme.spacing(1.5),
      marginLeft: theme.spacing(3),
    },
  })
);

type SearchBoxProps = {
  searchTerm: string;
  handleSearchTermChange: (event: ChangeEvent<HTMLInputElement>) => void;
  handleSearchButtonClick: () => void;
};

const SearchBox = ({
  searchTerm,
  handleSearchTermChange,
  handleSearchButtonClick,
}: SearchBoxProps) => {
  const classes = useStyles();
  return (
    <>
      <Typography variant="h6" component="h2">
        Search Your Favorite Movies!!
      </Typography>
      <Grid container>
        <Grid item xs={10}>
          <TextField
            id="search-movie"
            label="Search Movies!!"
            type="search"
            fullWidth
            value={searchTerm}
            onChange={handleSearchTermChange}
          />
        </Grid>
        <Grid item xs={2}>
          <Button
            variant="contained"
            color="primary"
            className={classes.searchButton}
            onClick={handleSearchButtonClick}
          >
            Search
          </Button>
        </Grid>
      </Grid>
    </>
  );
};

export default SearchBox;
