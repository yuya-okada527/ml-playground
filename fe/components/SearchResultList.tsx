import React from "react";
import {
  createStyles,
  List,
  ListItem,
  makeStyles,
  Theme,
  Typography,
} from "@material-ui/core";
import Link from "next/link";
import { Movie } from "../interfaces/index";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    movieCardTitle: {
      margin: theme.spacing(1),
    },
    underSearch: {
      marginBottom: theme.spacing(2),
    },
  })
);

type SearchResultProps = {
  movies: Movie[];
  searchedTerm: string;
};

const SearchResultList = ({ movies, searchedTerm }: SearchResultProps) => {
  const classes = useStyles();
  console.log(movies);
  return (
    <>
      <div className={classes.underSearch} />
      <Typography>Results for "{searchedTerm}"</Typography>
      <List>
        {movies.map((movie: Movie) => (
          <Link href={`/movies/${movie.movie_id}`} key={movie.movie_id}>
            <a>
              <ListItem key={movie.movie_id}>
                <Typography className={classes.movieCardTitle} variant="h6">
                  {movie.japanese_title
                    ? movie.japanese_title
                    : movie.original_title}
                  {movie.release_year ? " (" + movie.release_year + ") " : ""}
                </Typography>
              </ListItem>
              <hr />
            </a>
          </Link>
        ))}
      </List>
    </>
  );
};

export default SearchResultList;
