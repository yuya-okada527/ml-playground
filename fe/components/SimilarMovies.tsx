import React from "react";
import {
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
} from "@material-ui/core";
import { Movie } from "../interfaces";

type SimilarMoviesProps = {
  similarMovies: Movie[];
};

const SimilarMovies = ({ similarMovies }: SimilarMoviesProps) => (
  <>
    <Typography variant="h6">Similar Movies</Typography>
    <TableContainer component={Paper}>
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Rank</TableCell>
            <TableCell>Title</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {similarMovies.map((movie: Movie, index: number) => (
            <TableRow key={movie.movie_id}>
              <TableCell>{index + 1}</TableCell>
              <TableCell>
                {movie.japanese_title}{" "}
                {movie.release_year ? " (" + movie.release_year + ") " : ""}
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  </>
);

export default SimilarMovies;
