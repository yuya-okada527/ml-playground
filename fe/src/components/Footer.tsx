import {
  createStyles,
  List,
  ListItem,
  makeStyles,
  Theme,
} from "@material-ui/core";
import React from "react";

const MAIL_ADDRESS = "yuya.okada527@gmail.com";
const GITHUB_URL = "https://github.com/yuya-okada527/ml-playground";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    footer: {
      width: "100%",
      height: "65px",
      borderTop: "1px solid black",
      marginTop: theme.spacing(2),
    },
    footerList: {
      paddingTop: theme.spacing(0.5),
    },
    footerListItem: {
      paddingTop: theme.spacing(0.5),
      paddingBottom: theme.spacing(0),
    },
  })
);

/**
 * フッターを描画する
 */
const Footer: React.FC = () => {
  const classes = useStyles();
  return (
    <footer className={classes.footer}>
      <List className={classes.footerList}>
        <ListItem className={classes.footerListItem}>
          Contact: {MAIL_ADDRESS}
        </ListItem>
        <ListItem className={classes.footerListItem}>
          Github: {GITHUB_URL}
        </ListItem>
      </List>
    </footer>
  );
};

export default Footer;
