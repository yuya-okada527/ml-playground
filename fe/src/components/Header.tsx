import {
  AppBar,
  createStyles,
  makeStyles,
  Theme,
  Typography,
} from "@material-ui/core";
import React from "react";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    title: {
      margin: theme.spacing(2),
    },
  })
);

/**
 * ヘッダーを描画する
 */
const Header: React.FC = () => {
  const classes = useStyles();
  return (
    <header>
      <AppBar>
        <Typography variant="h4" component="h1" className={classes.title}>
          ML Playground
        </Typography>
      </AppBar>
    </header>
  );
};

export default Header;
