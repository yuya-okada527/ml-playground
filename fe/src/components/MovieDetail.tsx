import React from "react";
import {
  Card,
  CardMedia,
  CardContent,
  createStyles,
  makeStyles,
  Theme,
  Typography,
  Box,
} from "@material-ui/core";
import Image from "next/image";
import { Movie } from "../interfaces";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    cardRoot: {
      display: "flex",
      marginTop: theme.spacing(5),
      height: "450px",
      width: "98%",
      position: "relative",
    },
    posterImage: {
      flex: 1.2,
    },
    movieDescription: {
      flex: 2,
    },
    movieTitle: {
      marginBottom: theme.spacing(2),
    },
    providerInfo: {
      position: "absolute",
      bottom: 0,
      display: "flex",
      alignItems: "center",
    },
    providerIcon: {
      display: "inline",
    },
    providerComment: {
      fontSize: "13px",
      fontWeight: 300,
      marginRight: theme.spacing(1),
    },
  })
);

const MAX_OVERVIEW_LENGTH = 400;

const cutLongOverview = (overview?: string) => {
  if (!overview) {
    return overview;
  }

  if (overview.length > MAX_OVERVIEW_LENGTH) {
    return overview.substr(0, MAX_OVERVIEW_LENGTH) + "...";
  }

  return overview;
};

type MovieDetailProps = {
  movie_detail: Movie;
};

const MovieDetail: React.FC<MovieDetailProps> = ({ movie_detail }) => {
  const classes = useStyles();
  return (
    <Card className={classes.cardRoot}>
      <CardMedia
        className={classes.posterImage}
        image={movie_detail.poster_url}
      />
      <CardContent className={classes.movieDescription}>
        <Typography variant="h6" className={classes.movieTitle}>
          {movie_detail.japanese_title}
          {movie_detail.release_year
            ? " (" + movie_detail.release_year + ") "
            : ""}
        </Typography>
        <Typography variant="body1">
          {cutLongOverview(movie_detail.overview)}
        </Typography>
        <Box className={classes.providerInfo}>
          <Typography variant="subtitle2" className={classes.providerComment}>
            This product uses the TMDb API but is not endorsed or certified by
            TMDb.
          </Typography>
          <Image
            src="https://www.themoviedb.org/assets/2/v4/logos/v2/blue_square_1-5bdc75aaebeb75dc7ae79426ddd9be3b2be1e342510f8202baf6bffa71d7f5c4.svg"
            alt="Logo of Tmdb"
            width={40}
            height={40}
            className={classes.providerIcon}
          />
        </Box>
      </CardContent>
    </Card>
  );
};

export default MovieDetail;
