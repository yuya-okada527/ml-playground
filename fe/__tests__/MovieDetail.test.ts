import { makeOverview, makeTitle } from "../src/components/MovieDetail";

describe("テスト: makeOverview", () => {
  it("シナリオが空", () => {
    // テストデータ
    const overview = "";

    // 検証
    expect(makeOverview(overview)).toBe("");
  });

  it("シナリオがnull", () => {
    // テストデータ
    let overview;

    // 検証
    expect(makeOverview(overview)).toBe("");
  });

  it("シナリオの文字数 < 400", () => {
    // テストデータ
    const overview = "1234567890";

    // 検証
    expect(makeOverview(overview)).toBe(overview);
  });

  it("シナリオの文字列 > max_length", () => {
    // テストデータ
    const overview = "1234567890";
    const max_length = 5;

    // 検証
    const actual = makeOverview(overview, max_length);
    const expected = "12345...";
    expect(actual).toBe(expected);
  });
});

describe("テスト: makeTitle", () => {
  it("タイトルが無い場合", () => {
    // テストデータ
    const movie = emptyMovie();

    // 検証
    const actual = makeTitle(movie);
    const expected = "申し訳ありません。タイトルが表示できません。";
    expect(actual).toBe(expected);
  });

  it("日本語タイトルあり オリジナルタイトルなし 公開年なし", () => {
    // テストデータ
    const movie = emptyMovie();
    movie.japanese_title = "日本語タイトル";

    // 検証
    expect(makeTitle(movie)).toBe(movie.japanese_title);
  });

  it("日本語タイトルあり オリジナルタイトルあり 公開年あり", () => {
    // テストデータ
    const movie = emptyMovie();
    movie.japanese_title = "日本語タイトル";
    movie.original_title = "original title";
    movie.release_year = 2020;

    // 検証
    const actual = makeTitle(movie);
    const expected = "日本語タイトル (2020)";
    expect(actual).toBe(expected);
  });

  it("日本語タイトルなし オリジナルタイトルあり 公開年なし", () => {
    // テストデータ
    const movie = emptyMovie();
    movie.original_title = "original title";

    // 検証
    expect(makeTitle(movie)).toBe(movie.original_title);
  });

  it("日本語タイトルなし オリジナルタイトルあり 公開年あり", () => {
    // テストデータ
    const movie = emptyMovie();
    movie.original_title = "original title";
    movie.release_year = 2020;

    // 検証
    const actual = makeTitle(movie);
    const expected = "original title (2020)";
    expect(actual).toBe(expected);
  });
});

const emptyMovie = () => {
  return {
    movie_id: 0,
    original_title: "",
    japanese_title: "",
    overview: "",
    tagline: "",
    poster_url: "",
    backdrop_url: "",
    popularity: 0,
    vote_average: 0,
    release_date: "",
    release_year: 0,
    genre_labels: [],
    genres: [],
  };
};
