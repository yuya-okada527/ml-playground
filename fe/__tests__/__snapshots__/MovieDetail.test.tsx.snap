// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetail Snapshot Test 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-cardRoot-1"
>
  <WithStyles(ForwardRef(CardMedia))
    className="makeStyles-posterImage-2"
    image="https://image_url.png"
  />
  <WithStyles(ForwardRef(CardContent))
    className="makeStyles-movieDescription-3"
  >
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-movieTitle-4"
      variant="h6"
    >
      タイトル
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      シナリオ
    </WithStyles(ForwardRef(Typography))>
    <ProviderInfo />
  </WithStyles(ForwardRef(CardContent))>
</WithStyles(ForwardRef(Card))>
`;
